source/bolts/experimental/refraction.d

```diff
@@ -0,0 +1,723 @@
+    parameters: refractParameterList!(fun, localName),
+    udas: __traits(getAttributes, fun)
+    .length.iota.map!(
+        formatIndex!("@(__traits(getAttributes, %s)[%%d])".format(localName))).array,
+    attributes: functionAttributes!(fun),
+    static_: __traits(isStaticFunction, fun) && isAggregate!(__traits(parent, fun)),
+    body_: ";",
+    };
+
+    return model;
+ }
+
+///
+unittest {
+    pure @nogc int answer(lazy string question);
+    alias F = answer; // typically F is a template argument
+    static assert(
+        refract!(F, "F").mixture ==
+        "pure @nogc @system bolts.experimental.refraction.ReturnType!(F) answer(lazy bolts.experimental.refraction.Parameters!(F)[0] _0);");
+}
+
+///
+unittest {
+    import std.format;
+    import std.traits : FunctionAttribute;
+    import bolts.experimental.refraction;
+
+    interface GrandTour {
+        pure int foo() immutable;
+        @nogc @trusted nothrow ref int foo(out real, return ref int, lazy int) const;
+        @safe shared scope void bar(scope Object);
+    }
+
+    class Mock(Interface) : Interface {
+        static foreach (member; __traits(allMembers, Interface)) {
+            static foreach (fun; __traits(getOverloads, Interface, member)) {
+                mixin({
+                        enum Model = refract!(fun, "fun");
+                        if (is(ReturnType!fun == void)) {
+                            return Model.withBody("{}").mixture;
+                        } else if (Model.attributes & FunctionAttribute.ref_) {
+                            return Model.withBody(q{{
+                                        static %s rv;
+                                        return rv;
+                                    }}.format(Model.returnType)).mixture;
+                        } else {
+                            return Model.withBody(q{{
+                                        return %s.init;
+                                    }}.format(Model.returnType)).mixture;
+                        }
+                    }());
+            }
+        }
+    }
+
+    GrandTour mock = new Mock!GrandTour;
+    real x;
+    int i, l;
+    mock.foo(x, i, l++) = 1;
+    assert(mock.foo(x, i, l++) == 1);
+    assert(l == 0);
+}
+
+private enum isFunctionContainer(alias T) = is(T == module) || is(T == struct)
+    || is(T == class) || is(T == interface) || is(T == union);
+
+/**
+   Return a `Function` object refracting the function overload identified by
+   `name` and `index` in `Scope`, using the value of `localName` to represent
+   `Scope`. Store `index` in the `index` property of the Function.
+
+   Params:
+   Scope = a struct, class, interface, or union
+   localName = a string mixin that represents `Scope`
+   name = the name of the function
+   index = the index of the function in the set of overloaded function called `name`
+*/
+
+Function refract(alias Scope, string localName, string name, uint index)()
```

Comment:
Have you found use for this overload of `refract` in your use cases (aka do you need this for open methods or any other library?)

Reply:
> Have you found use for this overload of `refract` in your use cases (aka do you need this for open methods or any other library?)

My ducktyping lib uses the 3rd overload:

```d

  static auto makePointer(Function fun)
  {
    auto name = fun.name;
    return fun
      .removeMutabilityAttributes
      .insertParameters(
        0, Parameter(
          fun.copyMutabilityAttribute("void*")))
      .setName("function")
      .setBody("")
      .mixture ~ " %s%d;".format(name, fun.index);
  }
    
  struct Vtbl
  {
    mixin(refract!(Interface, "Interface").map!makePointer.join("\n"));
  }
```

This shows that in practice, the index is often needed - here it is to make unique identifiers for pointers to overloaded functions.

Currently I don't use overload \#2 but I think I did at one point, and I think that it will be needed at times, because the only way to form a symbolic reference to a function involves the module, the name and the index in the overload set.

Probably a `FunctionRef(alias Module, string name, int index)` would be a useful addition to bolts, along with a variant that takes a module, a name, and a signature, e.g.:

```d
module h2g2;

struct FunctionRef(alias _module, string _name, int _index) {
    alias Module = _module;
    enum name = _name;
    enum index = _index;
    alias Alias = __traits(getOverloads, Module, name)[index];
}

struct FunctionRef(alias _module, string _name, Pattern) {
    static foreach (member; __traits(allMembers, _module)) {
        static if (member == _name) {
            static foreach (i, fun; __traits(getOverloads, _module, _name)) {
                static if (!__traits(compiles, { enum _ = index; })) {
                    static if (is(typeof(&fun) == typeof(Pattern()))) {
                        alias Module = _module;
                        enum name = _name;
                        enum index = i;
                        alias Alias = fun;
                    }
                }
            }
        }
    }
}

string answer() { return "forty-two"; }
int answer() { return 42; }

alias f0 = FunctionRef!(h2g2, "answer", 0);
static assert (f0.Alias() == "forty-two");

alias f1 = FunctionRef!(h2g2, "answer", int function());
static assert (f1.Alias() == 42);
static assert (f1.index == 1);
```

Well, all that to say, addressing a function in presence of overloads is messy, and that's why I speculate that `refract(alias Scope, string localName, string name, uint index)()` will be useful in some cases, even though I don't use it anywhere right now (or maybe in a branch where I try to make openmethods work with templates, and that is a headache).

Reply:
> > static if (!__traits(compiles, { enum _ = index; })) ...
> 
> Huh ... neat trick!
> 
> I'm not sure that `FunctionRef!(h2g2, "answer", 0);` is better than `__traits(getOverloads, h2g2, "answer")[0];` but that pattern one is cool, is it practical?

I think that it's more readable, also we would add `static assert`s to better report cases where the function does not exist, or the overload index is out of range. The patterned version, I used in some of my many failed attempts of blending open methods with templates. But I have seen the question "how do I create an alias to an overloaded function" come up several times in the forums.

Reply:
> On the overloads. I feel a bit uneasy about the third. I'll try and explain:
> 
> The name of the utility is refraction and it makes sense when you want to perform refraction on a "thing". With the first overload this thing is a function. One the second overload, it's a function with in an overload set inside a scope. Which I also guess would be needed

It turns out that neither openmethods nor ducktyping use this overload. I'm fine with removing (or privatizing) it.

> but for the third I'm having trouble as you're not refacting the interface. You're returning an array of the interface's member functions that match a certain pattern. It's a get+filter.

Yeah maybe `refract` is getting a bit overloaded, semantically.

I considered the possibility that we add aggregate or module refraction later (actually a previous version, that used templates instead of ctfe objects, had it). Perhaps we should stick to the principle that `refract` always returns an object representing its first argument. We could have:

```d
module mymodule;
enum mod = refract!mymodule; // type: Module!mod
enum modFunctions = mod.functions!(filter); // type: Function[]
interface Foo { ... }
enum interfaceFunctions = refract!Foo.functions; // type: Function[]
```

In the meantime, it turns out that refracting all the functions in an aggregate is very useful. ducktyping does it in several places. Thinking about it, it would be useful if we could tell it to store a flat function index or an overload index. 

In the light of this, I suggest two options: 
- I rename overload \#3 to `extractFunctionsFrom` or (IMO better) `refractFunctionsFrom` (and remove \#2)
- I write a minimal implementation of module and aggregate refraction, just what we need to host the en-masse refraction of functions at this point (and we'll beef them up later).

What do you think?

Reply:
> But I have seen the question "how do I create an alias to an overloaded function" come up several times in the forums.

Yes I believe I have as well.

Something in the heart of bolts would be appropriate here indeed I believe. `AliasOf`? or should `std.meta.Alias` be extended?

Reply:
If bolts is to Phobos what Boost is to the C++ std lib, let's put it there, so we can showcase it, then try to push it to Phobos.

What about `bolts.members.overloadAlias`?